src/constraint_system/standard/composer.rs

```diff
@@ -587,31 +613,50 @@ impl StandardComposer {
         self.q_m.push(q_m);
         self.q_o.push(q_o);
         self.q_c.push(q_c);
-        self.q_4.push(Scalar::zero());
+        self.q_4.push(q_4);
 
         self.public_inputs.push(pi);
 
-        self.perm
-            .add_variables_to_map(a, b, c, self.zero_var, self.n);
+        self.perm.add_variables_to_map(a, b, c, d, self.n);
 
         self.n = self.n + 1;
 
         c
     }
+
     // q_m * a * b - c = 0
     pub fn mul(&mut self, q_m: Scalar, a: Variable, b: Variable, pi: Scalar) -> Variable {
+        self.big_mul(q_m, a, b, (Scalar::zero(), self.zero_var), pi)
+    }
+
+    // q_m * a * b + (q_4 * d) - c = 0
+    // returns c
+    pub fn big_mul(
+        &mut self,
+        q_m: Scalar,
+        a: Variable,
+        b: Variable,
+        q_4_d: (Scalar, Variable),
+        pi: Scalar,
+    ) -> Variable {
         let q_o = -Scalar::one();
         let q_c = Scalar::zero();
 
+        let q_4 = q_4_d.0;
+        let d = q_4_d.1;
+
         // Compute output wire
         let a_eval = self.variables[&a];
         let b_eval = self.variables[&b];
-        let c_eval = (q_m * a_eval * b_eval) + pi;
+        let d_eval = self.variables[&d];
+        let c_eval = (q_m * a_eval * b_eval) + (q_4 * d_eval) + pi;
```

Comment:
This is wrong imo. You would need to multiply the result by `-q_o` as I did for the `zexe_prover` branch to solve the errors on the implementation of the helpers.

It looked like this for mul :
```rust
        let o = (((self.eval(&l) * &self.eval(&r)) * &q_m) + &q_c + &pi) * &-q_o;
```

Check it here if needed: 585d281e

Also, this probably means that the previous merge of the addition implementation is wrong. We should fix #150 .
I can take it!

Reply:
I think this is working as intended;

- We do not need to scale by q_o because c is being computed.

- I also removed the constants because you can constrain a variable to a constant instead. This is a tradeoff, I think is acceptable because the `mul` and `add` functions are there for mostly simplicity.

Reply:
What I'm saying is that if you introduce a `q_o` > 1, the gate will cause the circuit to fail.

Remember that `o` should be equal to the rest, and in order to satisfy the equation, `o` has to consider the `q_o` factor, which means that `q_o` has to be multiplied to the rest.

To sum up, **c computation is not correct, to compute it correctly you should multiply `c_eval` by `-q_o`**

Reply:
In that function q_O is set at 1

Reply:
Since you're returning the value (which means that you're computing it inside of the function) you need to return the correct one. 

And without multiplying it by `-q_o` you're not doing it.
Did you check 585d281 ?
We can have a quick call to resolve this if needed!

Reply:
If you check the function, q_O = -1 , so multiplying by -q_O is the same as multiplying by 1.

I have added tests for this, you can check those to see

Reply:
If you try 2 it will fail. That's the problem.

Reply:
The user is not allowed to try 2. q_O is defined inside of the function.

Reply:
I was wrong, we can still scale the results.
Nevermind me then..